packages/hardhat/contracts/Pools.sol

```diff
@@ -0,0 +1,161 @@
+    uint256 public immutable excessUtilisationRate;
+    // Base variable borrow rate when Utilization rate = 0.
+    uint256 internal immutable baseBorrowRate;
+    // Slope of the variable interest curve when utilization rate > 0 and <= optimalUtilisationRate.
+    uint256 internal immutable slope1;
+    // Slope of the variable interest curve when utilization rate > optimalUtilisationRate.
+    uint256 internal immutable slope2;
+
+    mapping(address => Pool) private pools;
+
+    constructor (
+        uint256 _optimalUtilisationRate,
+        uint256 _baseBorrowRate,
+        uint256 _slope1,
+        uint256 _slope2
+    ) {
+        optimalUtilisationRate = _optimalUtilisationRate;
+        excessUtilisationRate = PRBMathUD60x18.SCALE - _optimalUtilisationRate;
+        baseBorrowRate = _baseBorrowRate;
+        slope1 = _slope1;
+        slope2 = _slope2;
+    }
+
+    function balance(address token) external view returns (uint256) {
+      return pools[token].balance;
+    }
+
+    function borrowed(address token) external view returns (uint256) {
+      return pools[token].borrowed;
+    }
+
+    function available(address token) external view override returns (uint qty) {
+        qty = pools[token].balance - pools[token].borrowed;
+    }
+
+    function balanceOf(address token, address owner) external view returns (uint) {
+        Pool storage pool = pools[token];
+        return pool.totalShare > 0
+        ? PRBMath.mulDiv(pool.balance, pool.shares[owner], pool.totalShare)
+        : 0;
+    }
+
+    function shareOf(address token, address owner) external view returns (uint) {
+        Pool storage pool = pools[token];
+        return pool.totalShare > 0
+        ? PRBMath.mulDiv(100, pool.shares[owner], pool.totalShare)
+        : 0;
+    }
+
+    function deposit(address token, uint256 amount) external validUAmount(amount) returns (uint256 share) {
+
+        Pool storage pool = pools[token];
+
+        share = pool.totalShare > 0
+        ? PRBMath.mulDiv(amount, pool.totalShare, pool.balance)
+        : amount * (10 ** ERC20(token).decimals());
+
+        pool.balance = pool.balance + amount;
+        pool.totalShare = pool.totalShare + share;
+        pool.shares[msg.sender] = pool.shares[msg.sender] + share;
+
+        ERC20(token).safeTransferFrom(msg.sender, address(this), amount);
+
+    }
+
+    function withdraw(address token, uint amount) external validUAmount(amount) {
+        Pool storage pool = pools[token];
+
+        require(pool.balance >= amount, "Amount too big");
+
+        uint256 share = PRBMath.mulDiv(pool.totalShare, amount, pool.balance);
+
+        require(pool.shares[msg.sender] >= share, "Amount gt than balance");
+
+        pool.balance = pool.balance - amount;
+        pool.totalShare = pool.totalShare - share;
+        pool.shares[msg.sender] = pool.shares[msg.sender] - share;
+
+        ERC20(token).safeTransfer(msg.sender, amount);
```

Comment:
I've read that, as a good practice, we should always call any 3rd party contract first, and then write to storage